### ISSUE TITLE
Interpolation bug at sample loop points

### ISSUE THREAD

**Issue by ghost** [schismtracker/schismtracker#27]
In this song: ftp://ftp.cubic.org/pub/audiotorium/1995/lasse.zip starting in pattern 2, there is an instrument (a bass) that sounds distorted. ModPlug and XMP play it right. MikMod doesn't play it totally right, but almost (it plays other parts wrongly).

Sorry, I don't know if it's the sample or the instrument. The pattern uses effect E40 (pitch down by 40) for the instrument to behave as a bass.

Tested with d7594cf (current HEAD at the time of writing).

I otherwise love the audio quality of Schism for mod-to-wav conversion. Great job!

**Comment by jangler** [comment]
Yeah, hm, Schism is interpolating that sample differently than Impulse and Milky do. Interpolation is usually not considered important enough to be considered a "playback compatibility issue" in players, but in this case it's quite noticeable and there's no way to get around it.

IIRC there might be some code that opts out of interpolating "chip" samples like this one that are under a certain length (in samples). Probably there should be a way to disable that.

**Comment by ghost** [comment]
Thanks for the prompt response. For what it's worth, changing the interpolation mode or the ramp setting in the Driver screen made no difference, which makes me think you're talking about interpolation happening at a different point.

**Comment by HybridDog** [comment]
it's sample 1 in openmpt:
![sample](https://cloud.githubusercontent.com/assets/3192173/17053449/540f619c-5002-11e6-8465-68ccb723f006.png)
the sample is obviously overmodulated

l used audacity to record it when it gets played by openmpt and schism
from openmpt:
![openmpt](https://cloud.githubusercontent.com/assets/3192173/17053506/a14d09dc-5002-11e6-96ee-a638f70d70c1.png)

schism with linear interpolation:
![schismlin](https://cloud.githubusercontent.com/assets/3192173/17053525/bf027e94-5002-11e6-9d62-b9cb1c8da5cd.png)

schism with the 8-Tap FIR Filter:
![schismfir](https://cloud.githubusercontent.com/assets/3192173/17053542/cd47ac9a-5002-11e6-8582-97ebe64f89cb.png)

l don't like the fir filter because the bright sounds become so quiet.

there's the sound from schism with linear interpolation:
https://github.com/HybridDog/miscellaneous/raw/master/sounds/sample1.flac
note that for any reason the recorded sound in audacity is slightly slower than the original

There's another problem: Often there are two speakers for the bright sound and one for the bass (sub woofer), and, as you can see in the audacity recordings, it's a dark stereo sound.
So because there's just one bass speaker, the dark sound is converted to mono for it, i.e. things like that nyancat 404 melody aren't played correctly: http://www.nyan.cat/music/404.ogg

**Comment by ghost** [comment]
HybridDog has inspired me to do some more research with a new angle. I think I see at least two issues here.

One is Schism's handling of the looping from the last sample to the first sample of the loop, and how that's used for interpolation. I think it's missing that the last sample is supposed to loop back to the first sample when interpolating. The last sample doesn't stand on its own, yet Schism seems to behave as if it did. I've marked it with a red **1** in the snapshot. That's probably the cause for the lack of cleanliness in the sound.

The other is the handling of the <s>end of the sample</s> [edit: I mean the wave output at note-off time]: a clear cut, marked with a red **2**. That causes a click at the end.

![lasse-schism](https://cloud.githubusercontent.com/assets/4267396/17062369/fad8328a-5032-11e6-9bec-c12769707f4b.png)

Compare with Milky:

![lasse-milky](https://cloud.githubusercontent.com/assets/4267396/17062503/7a42e5b0-5033-11e6-8ee7-8861e988f905.png)

This is how Milky draws the sample:
![milky-sample-view](https://cloud.githubusercontent.com/assets/4267396/17063069/dff1ba7e-5035-11e6-9110-0079279df27c.png)

It seems like Schism takes that interpretation literally, most notably the horizontal line at the end, judging by the output shown in Audacity. However, looking at the sample data (as saved to _wav_ by Milky), I see no reason to interpret the sample that way. For looped samples, the sample data should be interpreted with the last sample chaining back to the first. Like this (drawn by hand using the previous snapshot):

![how-milky-should-show-it](https://cloud.githubusercontent.com/assets/4267396/17063306/dab70068-5036-11e6-93fd-a3e6c119d250.png)

**Comment by HybridDog** [comment]
Looks like only openmpt sets the end right.

How does schism make the sample if you disable "Ramp volume at the start of sample" and use the cubic spline mixing mode? (shift+F5)

Edit: Considering the look of the first picture, did you already use cubic spline?

**Comment by ghost** [comment]
Hi Pedro,

This is something I reported on the scdev board a long time ago (before the githubization of schismtracker dev) and after speaking with Saga Musix it turned out it'd be a fair amount of effort to fix - even though it's a very simple problem there didn't seem to be a particularly simple solution IIRC. To save CPU time (interpolation happens often and muchly) I believe Schism just interpolates each byte (word?) of the sample into the next one in memory, rather than calculating which byte to interpolate into - n.b. this will be correct for all bytes/words except the one immediately after a loop point so it only really affects the sound of short "chip" samples.

The scdev board is offline now so I can't find exactly what was said, so take all of this with a pinch of salt, but I did look at trying to patch things back then and couldn't come up with much - there's the obvious solution of just doing the extra calculations and interpolating the last sample into the first one, but with short looped samples that would apparently eat CPU like nobody's business.

IIRC Saga/OpenMPT's (proposed?) solution is (was) to temporarily rewrite the byte/word after the loop end with the byte/word after the loop start during playback, which lets the crude method used presently work without millions of extra calculations - however that's not something I really have time to try myself right now.

Sorry this isn't particularly constructive, just thought I'd let you know it's something I briefly looked into before. Other people might have better things to say about it than me! :)

**Comment by HybridDog** [comment]
Please add a setting to enable more precise, cpu intensive interpolation.
Today l listened to Mortimers_chipdisco.mod and noticed the clicks. Converting the file didn't help of course.

l doubt that proper interpolation causes too much cpu usage since converting the file works around 400 times as fast as playing it:

```
$ time schism --diskwrite tmp.wav Mortimers_chipdisko.mod 
Diskwrite complete, exiting...

real    0m0.401s
user    0m0.192s
sys 0m0.064s
```

```
$ avprobe tmp.wav 
avprobe version 9.18-6:9.18-0ubuntu0.14.04.1, Copyright (c) 2007-2014 the Libav developers
  built on Mar 16 2015 13:19:10 with gcc 4.8 (Ubuntu 4.8.2-19ubuntu1)
[wav @ 0x1c5c380] max_analyze_duration reached
Input #0, wav, from 'tmp.wav':
  Duration: 00:02:38.40, bitrate: 1411 kb/s
    Stream #0.0: Audio: pcm_s16le, 44100 Hz, 2 channels, s16, 1411 kb/s
# avprobe output


```

**Comment by ghost** [comment]
My understanding is that each line of code added would end up being run hundreds of thousands of times per second per channel, I don't really know whether that's doable on modern hardware, but when I spoke to Saga Musix (a far more talented programmer than I'll ever be) about it before, he did suggest it wouldn't be worth doing. It might be better to just pick Saga's brain a bit and try to implement the OpenMPT fix, though of course one of the joys of open source is that we can all do pretty much whatever we want.

I simply don't have the time right now sadly.

**Comment by jangler** [comment]
If we're going to fix this, it's going to be by the OpenMPT hack solution referenced above – which sort of gives me the jeebies, but if it works it works.

_However_, let me be clear that @mb14's statements were _not_ an invitation to debate over the importance and priority of this issue. As maintainers of this software, we always appreciate a well-written bug report, but ultimately it's our call what gets changed, since we're the ones writing the code and dealing with the consequences. Of course, we always appreciate a well-written, well-tested, and scope-respectful pull request as well ;)

**Comment by ghost** [comment]
@HybridDog 

> How does schism make the sample if you disable "Ramp volume at the start of sample" and use the cubic spline mixing mode? (shift+F5)

Actually, I had it disabled. Enabling "Ramp volume at the start of sample" cured issue 2, so that one is actually not an issue.

I was using 8-tap FIR. Changing to cubic interpolation didn't change much.

**Comment by sagamusix** [comment]
As it has already been pointed out , "precise" interpolation would require per-sample modulo logic, which is _expensive_.

>  IIRC Saga/OpenMPT's (proposed?) solution is (was) to temporarily rewrite the byte/word after the loop end with the byte/word after the loop start during playback, which lets the crude method used presently work without millions of extra calculations - however that's not something I really have time to try myself right now.

This is actually what Schism and OpenMPT both did for a long time, but Schism removed this code at some point, and OpenMPT now uses a smarter solution. Don't ever rewrite the sample data directly, please.

These days, OpenMPT has a "loop wrap-around buffer", which is some extra data tacked at the end of the sample. It contains a certain amount of samples from before and after the loop end (at least as many as the interpolation filter looks forward/backward, i.e. 4 in the case of an 8-tap FIR filter, but preferrably a few more, maybe around 32-64 to reduce the switching overhead), and as soon as the sample play position gets close enough to the loop end, OpenMPT no longer interpolates the "normal" sample data but rather the tacked-on extra buffer. This way, it is easy to get the end of the loop to interpolate correctly - the start of the loop requires a little bit of extra work, but it's also doable (but also less nocitable if not done correctly - OpenMPT did not interpolate the loop start "correctly" until a few weeks ago).

Most of this can be implemented in or around get_sample_count (mixer.c), which is how I did it in OpenMPT.

**Comment by sagamusix** [comment]
Oh, and also:

> This is how Milky draws the sample:
> ...

Please don't ever take the output of sample editors at the sample start and end literally. It normally has nothing to do with how the loop would sound after interpolation. This is also true for OpenMPT.

> l doubt that proper interpolation causes too much cpu usage since converting the file works around 400 times as fast as playing it:

This is not how computing the complexity of a function works. It is slower by a large factor, and while it may not be obvious on your super-fast computer for a tiny 4-channel module (which you shouldn't play in Schism anyway, Schism is an IT tracker, not a MOD tracker), but on a slower computer or when playing a 64-channel module with NNA channels, things would get rather slow rather quickly. It is not possible to "simply insert a switch for precise playback", as it would require tons of changes in the mixer to allow for both the fast and the slow variant.

**Comment by ghost** [comment]
I like the idea of the loop wrap-around buffer. I was going to suggest to switch to the slow modulo-like method only near loop boundaries, but the buffer accomplishes the same result while being able to use the fast method, so that's preferable.

An alternative method would be: instead of a wrap-around buffer, have a note-off buffer. It is kind of an inside-out version. It would be a trade-off: it would be faster but it would use more memory.

**Comment by JosepMaJAZ** [comment]
Indeed. I also implemented the extra buffer on Psycle.  Methods Prefill, PreWork and PostWork
https://sourceforge.net/p/psycle/code/HEAD/tree/trunk/psycle/src/psycle/host/XMSampler.cpp#l172
The bunch of the change was in this commit, although it received fixes and corrections on successive commits: https://sourceforge.net/p/psycle/code/10482/
It even knows about ping-pong loops to setup two buffers and with different content. (That's why it's a bit more complicated than just having the extra samples). 
I no longer remember which song I used for testing, but it was a clear improvement to have this.

**Comment by BeRo1985** [comment]
When I was developing BeRoTracker, I addressed this issue by using sample copies. For each sample, I maintained the 100% original version without any loop modifications and created several loop-variant shadow copies. Each shadow copy had the loop border wraparounds fixed for individual loops. These shadow copies were used as the sample data source for the resampler routines, depending on which loop was active. While this approach did increase RAM usage, it was easy and quick to implement and ensured flawless loop playback. The sample outside the loops remained 100% original without any modifications. In my soundfont engine Sobanth, I employed a similar method but with a slightly more memory-optimized implementation.

**Comment by kay27** [comment]
Oh, that's true. Almost all of my modules composed in Impulse Tracker 2.14 and even Scream Tracker 3.21 experience issues with loops in samples when played in Schism Tracker. To fix them, I usually have to navigate to the sample and adjust the loop points by 1-2 ticks manually.

Interestingly, other players, such as Open Cubic Player, WinAMP, and Bass.DLL, play these modules correctly without any issues. Personally, I dislike the sound in ModPlug—perhaps it's due to its interpolation algorithm, though I’m not entirely sure.

I’ve attached an example module that demonstrates the issue. The first sample (01) has audible clicks at the beginning, which can be fixed by adjusting the loop start point from 3994 to 3992.

[tale.zip](https://github.com/user-attachments/files/17877677/tale.zip)

**Comment by mrpapersonic** [comment]
Wow I fixed this with a few commits almost a year ago(!) but never closed this issue.
